src/lightgbm/src/main/scala/TrainUtils.scala

```diff
@@ -36,95 +36,172 @@ private object TrainUtils extends Serializable {
-            result + " and is finished: " + isFinised)
-          iters = iters + 1
+  def translate(labelColumn: String, featuresColumn: String, weightColumn: Option[String],
+                validationData: Option[Array[Row]], log: Logger,
+                trainParams: TrainParams, inputRows: Iterator[Row]): Iterator[LightGBMBooster] = {
+    if (!inputRows.hasNext)
+      List[LightGBMBooster]().toIterator
+    val rows = inputRows.toArray
+    var trainDatasetPtr: Option[SWIGTYPE_p_void] = None
+    var validDatasetPtr: Option[SWIGTYPE_p_void] = None
+    try {
+      val (trainRows, validRows) =
+        if (validationData.isDefined) {
+          (rows, Some(validationData.get))
+        } else {
+          (rows, None)
         }
-        val bufferLength = LightGBMConstants.defaultBufferLength
-        val bufferLengthPtr = lightgbmlib.new_longp()
-        lightgbmlib.longp_assign(bufferLengthPtr, bufferLength)
-        val bufferLengthPtrInt64 = lightgbmlib.long_to_int64_t_ptr(bufferLengthPtr)
-        val bufferOutLengthPtr = lightgbmlib.new_int64_tp()
-        val tempM =
-          lightgbmlib.LGBM_BoosterSaveModelToStringSWIG(
-            boosterPtr.get, 0, -1, bufferLengthPtrInt64,
-            bufferOutLengthPtr)
-        val bufferOutLength = lightgbmlib.longp_value(lightgbmlib.int64_t_to_long_ptr(bufferOutLengthPtr))
-        // TODO: Move the reallocation logic inside the SWIG wrapper
-        val model =
-          if (bufferOutLength > bufferLength) {
-            lightgbmlib.LGBM_BoosterSaveModelToStringSWIG(
-              boosterPtr.get, 0, -1, bufferOutLengthPtr, bufferOutLengthPtr)
-          } else tempM
-        log.info("Buffer output length for model: " + bufferOutLength)
+      trainDatasetPtr = generateDataset(trainRows, labelColumn, featuresColumn, weightColumn, None)
+      if (validRows.isDefined && validRows.get.nonEmpty) {
+        validDatasetPtr = generateDataset(validRows.get, labelColumn, featuresColumn, weightColumn, trainDatasetPtr)
+      }
+      var boosterPtr: Option[SWIGTYPE_p_void] = None
+      try {
+        boosterPtr = createBooster(trainParams, trainDatasetPtr, validDatasetPtr)
+        trainCore(trainParams, boosterPtr, log, !validDatasetPtr.isEmpty)
+        val model = saveBoosterToString(boosterPtr, log)
         List[LightGBMBooster](new LightGBMBooster(model)).toIterator
       } finally {
         // Free booster
         if (boosterPtr.isDefined) {
-          LightGBMUtils.validate(lightgbmlib.LGBM_BoosterFree(boosterPtr.get),
-                                 "Finalize Booster")
+          LightGBMUtils.validate(lightgbmlib.LGBM_BoosterFree(boosterPtr.get), "Finalize Booster")
         }
       }
     } finally {
       // Free dataset
-      if (datasetPtr.isDefined) {
-        LightGBMUtils.validate(lightgbmlib.LGBM_DatasetFree(datasetPtr.get),
-                               "Finalize Dataset")
+      if (trainDatasetPtr.isDefined) {
```

Comment:
use option.foreach instead of if

Reply:
done!  fixed in some other places as well